src/batch_decoder.cpp

```diff
@@ -60,16 +65,16 @@ void batch_decoder::process(const int index)
     if (m_deterministic_mode)
         get_thread_local_random_engine() = m_random[index];
 
-    m_provider->provide(index, *m_inputs, *m_outputs);
+    m_provider->provide(index % m_batch_size, m_inputs->record(index), (*m_outputs)[index / m_batch_size]);
```

Comment:
reminder - please add a task about refactoring it a bit. ;)

Reply:
Yes, I did that